File: netpulse-web/src/features/observatory/attribution.test.ts
import { describe, expect, it } from "vitest";

import {
  incidentTitle,
  isCausalOverclaim,
} from "@/features/observatory/attribution";

describe("incident attribution", () => {
  it("prefers observed-failure language over ISP blame", () => {
    expect(
      incidentTitle({
        evidenceClass: "inferred_hypothesis",
        isolatedLayer: "isp",
        namedNetwork: "AS123",
      })
    ).toBe("Elevated connectivity failures observed");
    expect(isCausalOverclaim("ISP X caused the outage")).toBe(true);
  });

  it("may name a network only after measured isolation", () => {
    const title = incidentTitle({
      evidenceClass: "measured_fact",
      isolatedLayer: "isp",
      namedNetwork: "AS64500",
    });
    expect(title).toBe("Measured failures isolated to network AS64500");
    expect(isCausalOverclaim(title)).toBe(false);
  });
});

File: netpulse-api/internal/incidents/attribution_test.go
package incidents

import "testing"

func TestDefaultTitleDoesNotBlameAnISP(t *testing.T) {
	title := Title("inferred_hypothesis", "isp", "AS123")
	if title != "Elevated connectivity failures observed" {
		t.Fatalf("got %q", title)
	}
	if RejectsCausalOverclaim("ISP X caused the outage") == false {
		t.Fatal("causal overclaim must be flagged")
	}
}

func TestMeasuredIsolationMayNameANetworkWithoutSayingItCausedAnOutage(t *testing.T) {
	title := Title("measured_fact", "isp", "AS64500")
	if title != "Measured failures isolated to network AS64500" {
		t.Fatalf("got %q", title)
	}
	if RejectsCausalOverclaim(title) {
		t.Fatal("isolation language is not a causal overclaim")
	}
}
